js/ai/src/prompt.ts

```diff
@@ -14,79 +14,117 @@
+  z,
+} from '@genkit-ai/core';
+import { logger } from '@genkit-ai/core/logging';
 import { Registry } from '@genkit-ai/core/registry';
+import { toJsonSchema } from '@genkit-ai/core/schema';
+import { Message as DpMessage, PromptFunction } from 'dotprompt';
+import { existsSync, readdirSync, readFileSync } from 'fs';
+import { basename, join, resolve } from 'path';
 import { DocumentData } from './document.js';
 import {
+  generate,
   GenerateOptions,
   GenerateResponse,
+  generateStream,
   GenerateStreamResponse,
+  OutputOptions,
+  toGenerateRequest,
   ToolChoice,
 } from './generate.js';
+import { Message } from './message.js';
 import {
   GenerateRequest,
   GenerateRequestSchema,
   GenerateResponseChunkSchema,
+  GenerateResponseSchema,
+  MessageData,
+  ModelAction,
   ModelArgument,
   ModelMiddleware,
+  ModelReference,
+  Part,
 } from './model.js';
+import { getCurrentSession, Session } from './session.js';
 import { ToolAction, ToolArgument } from './tool.js';
 
-/**
- * Prompt implementation function signature.
- */
-export type PromptFn<
-  I extends z.ZodTypeAny = z.ZodTypeAny,
-  CustomOptionsSchema extends z.ZodTypeAny = z.ZodTypeAny,
-> = (input: z.infer<I>) => Promise<GenerateRequest<CustomOptionsSchema>>;
-
 /**
  * Prompt action.
  */
 export type PromptAction<I extends z.ZodTypeAny = z.ZodTypeAny> = Action<
   I,
   typeof GenerateRequestSchema,
-  typeof GenerateResponseChunkSchema
+  z.ZodNever
 > & {
   __action: {
     metadata: {
       type: 'prompt';
     };
   };
-  __config: PromptConfig;
+  __executablePrompt: ExecutablePrompt<I>;
 };
 
+/**
+ * Prompt action.
+ */
+export type ExecutablePromptAction<I extends z.ZodTypeAny = z.ZodTypeAny> =
+  Action<
+    I,
+    typeof GenerateResponseSchema,
+    typeof GenerateResponseChunkSchema
+  > & {
+    __action: {
+      metadata: {
+        type: 'executablePrompt';
+      };
+    };
+    __executablePrompt: ExecutablePrompt<I>;
+  };
+
 /**
  * Configuration for a prompt action.
  */
-export interface PromptConfig<I extends z.ZodTypeAny = z.ZodTypeAny> {
+export interface PromptConfig<
+  I extends z.ZodTypeAny = z.ZodTypeAny,
+  O extends z.ZodTypeAny = z.ZodTypeAny,
+  CustomOptions extends z.ZodTypeAny = z.ZodTypeAny,
+> {
   name: string;
+  variant?: string;
+  model?: ModelArgument<CustomOptions>;
+  config?: z.infer<CustomOptions>;
   description?: string;
-  inputSchema?: I;
-  inputJsonSchema?: JSONSchema7;
+  input?: {
+    schema?: I;
+    jsonSchema?: JSONSchema7;
+  };
+  system?: string | Part | Part[] | PartsGenerator<z.infer<I>>;
+  prompt?: string | Part | Part[] | PartsGenerator<z.infer<I>>;
+  messages?: string | MessageData[] | MessageGenerator<z.infer<I>>;
+  docs?: DocumentData[] | DocumentGenerator<z.infer<I>>;
+  output?: OutputOptions<O>;
+  maxTurns?: number;
+  returnToolRequests?: boolean;
   metadata?: Record<string, any>;
   tools?: ToolArgument[];
   toolChoice?: ToolChoice;
   use?: ModelMiddleware[];
 }
```

Comment:
Interfaces this big make me a little nervous...is there another canonical interface that we can derive this from to make it less likely to drift? Or should we somehow document an "if change" type thing?

Reply:
It tries to follow GenerateOptions, but with a lot of sugar... can probably extend from it, but would have to omit and override like half the fields...

Reply:
Eh, it's probably fine...